internal/tcp/tcp.go

```diff
@@ -228,349 +105,220 @@ var (
 
-	return packet.Marshal(), nil
-}
+	// gopacket serialization options for IP header are OS specific
+	optsIP := ip.GetIPLayerOptions()
 
-// Marshal emits raw bytes for the packet.
-func (pkt *tcpPacket) Marshal() []byte {
-
-	buf := new(bytes.Buffer)
-	binary.Write(buf, binary.BigEndian, pkt.srcPort)
-	binary.Write(buf, binary.BigEndian, pkt.dstPort)
-	binary.Write(buf, binary.BigEndian, pkt.seqNum)
-	binary.Write(buf, binary.BigEndian, pkt.ackNum)
-
-	var mix uint16
-	mix = uint16(pkt.dataOffset)<<12 | // top 4 bits
-		uint16(pkt.reserved)<<9 | // 3 bits
-		uint16(pkt.ECN)<<6 | // 3 bits
-		uint16(pkt.ctrl) // bottom 6 bits
-	binary.Write(buf, binary.BigEndian, mix)
-
-	binary.Write(buf, binary.BigEndian, pkt.window)
-	binary.Write(buf, binary.BigEndian, pkt.checksum)
-	binary.Write(buf, binary.BigEndian, pkt.urgent)
-
-	for _, option := range pkt.options {
-		binary.Write(buf, binary.BigEndian, option.Kind)
-		if option.Length > 1 {
-			binary.Write(buf, binary.BigEndian, option.Length)
-			binary.Write(buf, binary.BigEndian, option.Data)
+	// When replying with SYN+ACK, a time-stamped payload is included
+	if flags.syn != false && flags.ack != false {
+		payloadTime, err := timeNow().MarshalBinary()
+		if err != nil {
+			return nil, err
 		}
+		payloadLayer := gopacket.Payload(payloadTime)
+		payloadLayer.SerializeTo(buf, optsTCP)
 	}
 
-	if pkt.tcpPayload != (tcpPayload{}) {
-		byteTime, err := pkt.tcpPayload.Ts.MarshalBinary()
-		if err == nil {
-			binary.Write(buf, binary.BigEndian, byteTime)
-		}
+	tcpLayer := &layers.TCP{
+		SrcPort:    srcPort,
+		DstPort:    dstPort,
+		Seq:        seqNum,
+		Ack:        ackNum,
+		DataOffset: uint8(defines.TCPHeaderLength / 4), // TCP Header size in 32-bit words
+		SYN:        flags.syn,
+		RST:        flags.rst,
+		ACK:        flags.ack,
+		Window:     defines.TCPWindowSize,
+		Checksum:   0, // computed upon serialization
 	}
 
-	out := buf.Bytes()
+	// Length of TCP portion is payload length + fixed 20 bytes for Header
+	tcpLen := defines.TCPHeaderLength + len(buf.Bytes())
 
-	return out
-}
-
-func (tcp *tcpHeader) hasFlag(flagBit byte) bool {
-	return tcp.ctrl&flagBit != 0
-}
-
-// TCP Checksum
-func checksum(af string, data []byte, srcip, dstip *net.IP) (uint16, error) {
-
-	// the pseudo header used for TCP c-sum computation
-	var pseudoHeader []byte
-
-	pseudoHeader = append(pseudoHeader, *srcip...)
-	pseudoHeader = append(pseudoHeader, *dstip...)
-	switch af {
-	case "ip4":
-		pseudoHeader = append(pseudoHeader, []byte{
-			0,
-			6,                  // protocol number for TCP
-			0, byte(len(data)), // TCP length (16 bits), w/o pseudoheader
-		}...)
-	case "ip6":
-		pseudoHeader = append(pseudoHeader, []byte{
-			0, 0, 0, byte(len(data)), // TCP length (32 bits), w/0 pseudoheader
-			0, 0, 0,
-			6, // protocol number for TCP
-		}...)
-	default:
-		return 0, errors.New("unhandled AF family")
+	ipLayer, err := ip.GetIPHeaderLayer(af, dscpv, uint16(tcpLen), srcAddr, dstAddr)
```

Comment:
Can we compute this a bit higher above so that timeNow() payload gets measured as close to the actual transmission as possible?